conjure-java-core/src/test/resources/example-service.yml

```diff
@@ -224,6 +224,21 @@ services:
             type: optional<rid>
             param-type: query
 
+      testOptionalHeaderParam:
+        http: POST /test-optional-header-param
+        args:
+          query: string
+          maybeHeader:
+            type: optional<string>
+            param-type: header
+            param-id: MaybeHeader
+          implicit:
+            type: rid
+            param-type: query
+          maybeQuery:
+            type: optional<rid>
+            param-type: query
```

Comment:
This example is generated into the suffixed files, but those are never compiled -- it would be helpful to define this somewhere that outputs into a source set that's also compiled for additional verification, otherwise duplicate methods with the same signature won't be caught by CI.

Reply:
do you have a place in mind where I could put this example?

I think the cases where type clashes may cause the headers to not compile would be exactly when the _first_ query param is optional and has the same type as _any_ header param. This could be detected in code and worked around by not generating backcompat methods for header params if they would clash. It's likely a corner case in practice, but I appreciate that we need something that always compiles.

Reply:
Adding it to a new service defined in ete-service.yml should provide compilation verification. Using a new service instead of an existing one means you won't need to update any resource implementations to stub the method.

Is there danger in upgrading from an endpoint with an optional header to an endpoint with an optional header and optional query parameter that the same parameter index and type becomes used for another field?

Reply:
I guess to be safe you'd want there to always be a non-optional param (of any kind) directly after the last optional header param.